Topics/data generator

zakalawe opened the pull request:
Work in progress on the simulation generator. Reset data seems bogus, intervals are much too long, so probably I have misunderstood something in the data generation API, but currently it's escaping me!

zakalawe committed:
Settings data defined

This won’t build yet, since I didn’t include commented-out code
to compile the rest of the plugin.

zakalawe committed:
Basic CMake build system

Needs Linux, Windows support and installation path handling still.

zakalawe committed:
Data-generator work in progress

Marcus10110 commented on source/AsyncRgbLedSimulationDataGenerator.cpp:
```diff
@@ -1,11 +1,22 @@
 
 #include <AnalyzerHelpers.h>
 
+// these values are from the WS2811, in high-speed mode
```
This is fine for now, but later we'll want to replace this with a table of data that contains the details for all controllers. We'll also want to include in comments the URLs of the sources of this information.

Marcus10110 reviewed (changes requested):
Everything looks great! I did want to warn against using integer nanoseconds though for simulation. This introduces rounding error, which can produce errors in the simulated output.

This won't matter for decoding, but for simulating, we have a helper class called ClockGenerator which might be useful. It does have some limitations though, but it's good at accumulating rounding error. It's normally used for generating clocks, but it can accumulate the rounding error from any double precision length of time. Just ignore the target_frequency parameter, since we won't use it.

For instance, let's say you want to advance 600 ns, and the recording was taken at 12 MS/s. 
In as it stands, nSecPerSample should be 83.3333, but it will be rounded to 83.
Then the mSampleCounts entry should be 7.2, but will be stored as 7. Everytime you try to advance 600ns, you will instead advance 583ns, an error of 2.8%.

However the following avoids that:
ClockGenerator generator;
ClockGenerator.Init( 1.0, 12000000); //frequency isn't used here, so I'm using 1.0.
ClockGenerator.AdvanceByTimeS(600E-9);
The last call should return 7 most of the time, but occasionally return 8 when the internal error accumulator exceeds a complete sample.

ClockGenerator could be used to compute the number of samples to advance in all situations.

Each ClockGenerator instance keeps track of its own error, so you should try to reuse the same instance as much as possible.

zakalawe commented:
Ah I was aware of the integer issue, but hadn't understood I could use the clock generator to solve it, I'll make that change tomorrow. 

Regarding the simulation data, should I add a table so that the data matches the selected controller? That's no problem of course.

Marcus10110 commented:
@zakalawe, for the data table, we could actually put all unique information about a given controller into a config object of some sort, including high and low times for high and slow speed, (and a flag that indicates ih it supports high speed) but also include information like the color ordering (RGB vs BGR etc), bits per color, LEDs per module, string name, or any other detail that we might switch on later.

zakalawe committed:
Revised settings after discussion with Mark

zakalawe committed:
Data generation fixes

zakalawe committed:
Settings bug fix in timing read-out

zakalawe committed:
Settings MSVC float -> int fix

zakalawe committed:
MSVC / Linux Cmake build from Mark

zakalawe committed:
Add RGB helper struct

Use this to deal with different color layouts and bit sizes in
a cleaner way.

Marcus10110 commented on source/AsyncRgbLedAnalyzerSettings.cpp:
```diff
@@ -67,7 +107,46 @@ const char* AsyncRgbLedAnalyzerSettings::SaveSettings()
+U32 AsyncRgbLedAnalyzerSettings::DataTimeNSecHigh(BitState value) const
+{
+    const auto& d = mControllers.at(mLEDController);
+    return mIsHighSpeedMode ? d.mDataTimingHighSpeedNsec[value][0]
```
mIsHighSpeedMode probably should not be a member of the AsyncRgbLedAnalyzerSettings. Exactly how this works in analysis is yet to be seen, but I suggest we pass it in as a parameter, so the simulation data generator can use it for now.

zakalawe replied:
Ah yes, we discussed that, no problem.

Marcus10110 commented on source/AsyncRgbLedAnalyzerSettings.cpp:
```diff
@@ -67,7 +107,46 @@ const char* AsyncRgbLedAnalyzerSettings::SaveSettings()
+    return mControllers.at(mLEDController).mChannelCount;
+}
+
+U32 AsyncRgbLedAnalyzerSettings::DataTimeNSecHigh(BitState value) const
```
I suggest we use seconds and doubles instead of integers and nanoseconds. Thoughts?

zakalawe replied:
I was habitually using integers for performance reasons, but happy to use doubles and seconds, there's probably no difference.

Marcus10110 commented on source/AsyncRgbLedAnalyzerSettings.cpp:
```diff
@@ -67,7 +107,46 @@ const char* AsyncRgbLedAnalyzerSettings::SaveSettings()
+    return mControllers.at(mLEDController).mResetTimeNSec;
+}
+
+auto AsyncRgbLedAnalyzerSettings::GetColorLayout() const -> ColorLayout
```
sweet syntax, I haven't seen this in the wild before. Since we know the output type, is there benefit to using `auto` and `-> ColorLayout` over just using ColorLayout as the return type?

zakalawe replied:
This syntax is useful for class-scoped enums, but since I moved ColorLayout into the Helpers.h, we no longer even need this syntax, so I can remove it.

Marcus10110 commented on source/AsyncRgbLedAnalyzerSettings.h:
```diff
@@ -1,27 +1,92 @@
+        U8 mBitsPerChannel; // = 8;
+        U8 mChannelCount; // = 3;
+        U32 mResetTimeNSec;
+        U32 mDataTimingNSec[2][2];
```
I would suggest we expand this into 4 members, or creating a new struct to hold the contents. Right now this is a little confusing as to where the data goes.
Also, in light of the WS2813, I suggest we store the minimum, maximum, and nominal values for time high and time low, for both high speed and low speed.
This is because in the case of WS2813, the minimum low time is 300ns, and the maximum is 100us. The nominal is definitely not the halfway point, instead it should be 1.25us - nominal time high. Nominal values would just be used by simulation, not but the analysis part.

Maybe something like this:
```C++
struct LedTimingTolerance
{
    double mMinimumTimeS;
    double mNominalTimeS;
    double mMaximumTimeS;
};

struct LedControllerData
{
    //...
    LedTimingTolerance mHighSpeedOne;
    LedTimingTolerance mHighSpeedZero;
    LedTimingTolerance mLowSpeedOne;
    LedTimingTolerance mLowSpeedZero;
};
```

zakalawe replied:
Okay, I was wondering to store nominal and the tolerance, but the WS2813 values make that tricky as well. I'll adapt it base don what you propose here.

Marcus10110 commented on source/AsyncRgbLedSimulationDataGenerator.cpp:
```diff
@@ -15,57 +19,97 @@ AsyncRgbLedSimulationDataGenerator::~AsyncRgbLedSimulationDataGenerator()
+	mMaximumChannelValue = (1 << mSettings->BitSize()) - 1;
+
+    // TODO pass in the analyzer and call GetMinimumSampleRate?
+    mClockGenerator.Init( 12 * 1000000, mSimulationSampleRateHz );
```
this is very minor, but I suggest we pass in 1 as the frequency, for clarity, since it isn't used, and add a comment stating that the clock frequency is never used.

zakalawe replied:
No problem, will do.

Marcus10110 commented on source/AsyncRgbLedSimulationDataGenerator.cpp:
```diff
@@ -15,57 +19,97 @@ AsyncRgbLedSimulationDataGenerator::~AsyncRgbLedSimulationDataGenerator()
+
+	mLEDSimulationData.SetChannel( mSettings->mInputChannel );
+	mLEDSimulationData.SetSampleRate( simulation_sample_rate );
+	mLEDSimulationData.SetInitialBitState( BIT_HIGH );
```
I recommend we set the initial state low. elsewhere I'll mention that we should also remove the high pulse at the end of the data burst. The simulation should just begin as if it was already in a reset pulse.

zakalawe replied:
I am slightly confused by this, since in your earlier changes you set the initial bit state to low as well. My reading of the datasheets is that the the line sits high 'normally' and is pulled low to trigger a reset - that's also why I added the 'return to high' in the simulation after a data bit.

I guess your reading is that the line sits in the RESET state all the time, except when data is being sent, and I think what you wrote about changes to ReadBit might confirm that. Not a problem, it's just not what I understood from the timing diagrams in the data-sheets.

Marcus10110 replied:
In the implementations I've seen, I haven't seen the line left high. It definitely could be left high, especially if the controller was extra concerned with the exact time the reset event occurs. For our simulation, let's leave it low all the time though, to better match the implementations I've already seen.

Marcus10110 commented on source/AsyncRgbLedSimulationDataGenerator.cpp:
```diff
@@ -15,57 +19,97 @@ AsyncRgbLedSimulationDataGenerator::~AsyncRgbLedSimulationDataGenerator()
 
-	mSerialSimulationData.Transition();  //low-going edge for start bit
-	mSerialSimulationData.Advance( samples_per_bit );  //add start bit time
+	for (int i=0; i<3; ++i) {
```
Depending on how we want to handle decoding the TM1809 parts (9 channels per frame) we may want to loop here up to LEDChannelCount(). More on that later.

Marcus10110 replied:
I just noticed your comment above about being 6 or one or 2 of the other. ignore this comment, I think it's just as messy to either A) leave it as is, always writing 3 channels / one color at a time as B) having this function write either 1 or 3 colors depending on the LED controller type, and accept an array of color inputs. we can leave it as-is.

zakalawe replied:
Okay, I don't have any strong opinion here at all, just say if you have a preference.

Marcus10110 commented on source/AsyncRgbLedSimulationDataGenerator.cpp:
```diff
@@ -15,57 +19,97 @@ AsyncRgbLedSimulationDataGenerator::~AsyncRgbLedSimulationDataGenerator()
+	while( mLEDSimulationData.GetCurrentSampleNumber() < adjusted_largest_sample_requested )
 	{
-		CreateSerialByte();
+        WriteReset();
```
For testing purposes, we should toggle high speed mode every few cycles.
Since high speed mode isn't a setting of the analyzer, I recommend keeping it as a member variable of the simulation data generator, instead of a member of the settings class.

zakalawe replied:
Okay, I hadn't realised high-speed mode would vary like that, it's no problem to do that.

Marcus10110 replied:
Generally it won't in the real world, but it's a good way to exercise different features of the analyzer.

Marcus10110 commented on source/AsyncRgbLedSimulationDataGenerator.cpp:
```diff
@@ -15,57 +19,97 @@ AsyncRgbLedSimulationDataGenerator::~AsyncRgbLedSimulationDataGenerator()
+void AsyncRgbLedSimulationDataGenerator::WriteReset()
+{
+    // interstitial after last data bit, before the reset goes low
+    mLEDSimulationData.Advance( mClockGenerator.AdvanceByTimeS(1e-5) );
```
lets remove this pulse, the last bit should go right into the reset, as observed on real-world controllers. (see email with capture).

zakalawe replied:
Right, will address this with the ReadBit changes in the analysis code of course.

Marcus10110 commented on source/AsyncRgbLedSimulationDataGenerator.cpp:
```diff
@@ -15,57 +19,97 @@ AsyncRgbLedSimulationDataGenerator::~AsyncRgbLedSimulationDataGenerator()
+    // interstitial after last data bit, before the reset goes low
+    mLEDSimulationData.Advance( mClockGenerator.AdvanceByTimeS(1e-5) );
+    mLEDSimulationData.Transition(); // go low
+    // scale by 120% since reset time is the minimum I believe
```
I suggest each controller lists a minimum reset pulse time (seconds, double) and a nominal reset pulse time (seconds, double). The nominal value is used exclusively by the simulation data generator.
This is in the interest of making a great example. Every protocol analyzer will have to handle timing tolerance, and I would like to make ours very verbose and clear as to exactly how we handle these tolerances.

zakalawe replied:
Understood, will do.

Marcus10110 commented on source/AsyncRgbLedSimulationDataGenerator.cpp:
```diff
@@ -15,57 +19,97 @@ AsyncRgbLedSimulationDataGenerator::~AsyncRgbLedSimulationDataGenerator()
-		mSerialSimulationData.Advance( samples_per_bit );
+void AsyncRgbLedSimulationDataGenerator::WriteUIntData( U32 data, U8 bit_count )
+{
+	U32 mask =  1 << (bit_count - 1);
```
just checking here, have you confirmed that all controllers are MSB first? I haven't looked.

Marcus10110 replied:
Also, I AnalyzerHelpers.h contains a class `BitExtractor`, which we should use instead of manually shifting bits using a mask. `BitExtractor` is used for extracting bits from a word. The `DataBuilder` class is used for constructing a word from bits, which you should use in the analysis thread.

zakalawe replied:
I think they are all MSB from what I read, but I can check again.

I'll look at the BitExtractor and DataBuilder classes and use them as suggested.

Marcus10110 commented on source/AsyncRgbLedSimulationDataGenerator.cpp:
```diff
@@ -15,57 +19,97 @@ AsyncRgbLedSimulationDataGenerator::~AsyncRgbLedSimulationDataGenerator()
+void AsyncRgbLedSimulationDataGenerator::WriteBit(bool b)
+{
+    const BitState bs = b ? BIT_HIGH : BIT_LOW;
+    const double highSamples = mClockGenerator.AdvanceByTimeS( mSettings->DataTimeNSecHigh( bs ) * NSEC_TO_SEC );
```
`AdvanceByTimeS` returns a U32.
As mentioned above too, lets store the timing values in settings as double (seconds) instead of integer (nanoseconds).

zakalawe replied:
To be clear, when you mention the return type of AdvanceByTimeS, is that in the context of MSVC needing an explicit int -> double cast? Or something else?

Marcus10110 replied:
AdvanceByTimeS takes a floating point number of seconds and returns an integer number of samples. We should track all measures of time as doubles and all measures of samples as integers, this is generally how we do it internally too. We usually try to stay in samples as much as we can for performance. In this case though, we need to keep time in doubles the entire time so that we can benefit from the ClockGenerator's error accumulator.

Marcus10110 commented on source/AsyncRgbLedSimulationDataGenerator.cpp:
```diff
@@ -15,57 +19,97 @@ AsyncRgbLedSimulationDataGenerator::~AsyncRgbLedSimulationDataGenerator()
+    mLEDSimulationData.Advance( highSamples );
+	mLEDSimulationData.Transition(); // go low
+    mLEDSimulationData.Advance( lowSamples );
+	mLEDSimulationData.Transition(); // go high to end this bit
```
Right now, this function assumes that a transition has just been added to the simulation data before getting called, and it leaves a transition on the channel after getting called.

Since we're eliminating the positive pulse that you're currently adding at the beginning of the reset pulse function, I propose the following:
1. The initial bit state is low.
2. The reset function does not add any transitions, it just calls advance. (the function enters low and exits low)
3. This function, Writebit, transitions immediately on call (as it does) but it does not transition again after the final low pulse, leaving the channel low. (the function enters low and exits low)

This results in the mLEDSimulationData always being low outside of calls to WriteBit and WriteReset.
As an example analyzer, this is an important concept. If there were 3 or more functions that all created events on the bus, the developer would need to keep track of what the state is of the simulation channel at the entrance to each function. Keeping it low outside of all functions makes it easier to reason about.

zakalawe replied:
My implementation here was based on the assumption that the line is always high by default, and was taken low-then-high by all the Write functions (WriteBit or WriteReset). So I think the only change here (which I will do) is to change the convention that we presume data signal is low on entry and exit rather than high.

Marcus10110 replied:
perfect

Marcus10110 commented on source/AsyncRgbLedSimulationDataGenerator.h:
```diff
@@ -19,11 +22,20 @@ class AsyncRgbLedSimulationDataGenerator
+
+    void WriteRGBTriple( const RGBValue& rgb );
+	void WriteUIntData( U32 data, U8 bit_count );
+	void WriteBit(bool b);
```
let's change the signature from bool to BitState

zakalawe replied:
Will do.

Marcus10110 commented on source/AsyncRgbLedSimulationDataGenerator.cpp:
```diff
@@ -15,57 +19,97 @@ AsyncRgbLedSimulationDataGenerator::~AsyncRgbLedSimulationDataGenerator()
+}
 
-		mSerialSimulationData.Advance( samples_per_bit );
+void AsyncRgbLedSimulationDataGenerator::WriteUIntData( U32 data, U8 bit_count )
```
If we're explicitly holding onto the RGB values as U16s, we should probably write this function to accept a U16 instead of a U32, just for consistency.

zakalawe replied:
Will do.

Marcus10110 commented on source/AsyncRgbLedSimulationDataGenerator.h:
```diff
@@ -19,11 +22,20 @@ class AsyncRgbLedSimulationDataGenerator
 
+	// largest value for a color channel in the selected controller.
+	// this is 2^bitSize - 1
+	U32 mMaximumChannelValue = 255;
```
since the value set here is never used (it's over-written in Initialize) should we still initialize it here?

zakalawe replied:
This is just force of habit for me at this point, in case someone adds new constructors or init code paths and forget to initialise members. But as you it can be removed if you prefer.

Marcus10110 reviewed (changes requested):
This looks solid. I have a lot of very minor change requests. The main thing is removing the high pulse at the start of every reset pulse, I have detailed notes about that in the comments.
Sorry if these seem overly critical and trivial. I hope my feedback leads to a even cleaner example analyzer.

zakalawe committed:
Settings / data-generator changes based on review

- define timings with tolerances, min, max
- use BitExtractor in the data-generator
- assume the data signal defaults to low, not high
- define some helper structs for managing timings with tolerances
- generate both high-sped and low-speed data
